lib/ddtrace/span.rb

```diff
@@ -91,19 +95,44 @@ def set_error(e)
 
     # Mark the span finished at the current time and submit it.
     def finish(finish_time = nil)
+      # A span should not be finished twice. Note that this is not thread-safe,
+      # finish is called from multiple threads, a given span might be finished
+      # several times. Again, one should not do this, so this test is more a
+      # fallback to avoid very bad things and protect you in most common cases.
       return if finished?
 
-      @end_time = finish_time.nil? ? Time.now.utc : finish_time
-      @tracer.record(self) unless @tracer.nil?
+      # Provide a default start_time if unset, but this should have been set by start_span.
+      # Using now here causes 0-duration spans, still, this is expected, as we never
+      # explicitely say when it started.
+      @start_time ||= Time.now.utc
+
+      @end_time = finish_time.nil? ? Time.now.utc : finish_time # finish this
+
+      # Finish does not really do anything if the span is not bound to a tracer and a context.
+      return self if @tracer.nil? || @context.nil?
+
+      # spans without a service would be dropped, so here we provide a default.
+      # This should really never happen with integrations in contrib, as a default
+      # service is always set. It's only for custom instrumentation.
+      @service ||= @tracer.default_service unless @tracer.nil?
+
+      begin
+        @context.close_span(self)
+        @tracer.record(self)
+      rescue StandardError => e
+        Datadog::Tracer.log.debug("error recording finished trace: #{e}")
+      end
       self
     end
 
-    # Proxy function that flag a span as finished with the given
-    # timestamp. This function is used for retro-compatibility.
-    # DEPRECATED: remove this function in the next release
-    def finish_at(finish_time)
-      finish(finish_time)
-    end
+    # BREAKING CHANGE: finish_at has removed.
+    #
```

Comment:
Next step would be: remove this comment. But I would still leave it for this release, so that if someone was relying on it, it's just so obvious it disappeared (without browsing Github history).